File: app/src/main/java/com/SagaLogistic/backend/models/ItemImpl.java
package com.SagaLogistic.backend.models;

import android.os.Parcel;
import android.os.Parcelable;

public class ItemImpl implements Item{

    public static final Parcelable.Creator CREATOR = new Creator() {
        @Override
        public ItemImpl createFromParcel(Parcel in) {
            return new ItemImpl(in);
        }

        @Override
        public ItemImpl[] newArray(int size) {
            return new ItemImpl[size];
        }
    };

    private String key;
    private String name;
    private Float weight;

    public ItemImpl(String key, String name, Float weight){
        this.key = key;
        this.name = name;
        this.weight = weight;
    }

    public ItemImpl(String name, Float weight){
        this.name = name;
        this.weight = weight;
    }

    public String getKey() {
        return key;
    }

    public void setKey(String key) {
        this.key = key;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Float getWeight() {
        return weight;
    }

    public void setWeight(Float weight) {
        this.weight = weight;
    }

    public ItemImpl(Parcel in){
        this.key = in.readString();
        this.name = in.readString();
        this.weight = in.readFloat();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(this.key);
        dest.writeString(this.name);
        dest.writeFloat(this.weight);
    }
}

File: app/src/main/java/com/sagalogistics/backend/implementation/models/ItemImpl.kt
package com.sagalogistics.backend.implementation.models

import android.os.Parcelable
import com.google.firebase.database.Exclude
import android.os.Parcel
import android.os.Parcelable.Creator
import com.sagalogistics.backend.api.models.Item
import java.lang.IllegalArgumentException

class ItemImpl : Item, Parcelable {
    @get:Exclude
    @set:Exclude
    override var key: String? = null
    override lateinit var name: String
    override var weight: Float = 0f
        set(value) {
            if(value < 0)
                throw IllegalArgumentException("Weight can't be negative")
            field = value
        }
    override var upperVariance: Float = 0f
        set(value) {
            if(value < 0 || value > 100)
                throw IllegalArgumentException("Percentage must be between 0 and 100")
            field = value
        }
    override var lowerVariance: Float = 0f
        set(value) {
            if(value < 0 || value > 100)
                throw IllegalArgumentException("Percentage must be between 0 and 100")
            field = value
        }
    override var image: String? = null

    /**
     * Empty constructor needed for serialization; should not be used
     */
    constructor() {}

    /**
     * Standard constructor
     */
    constructor(key: String? = null, name: String, weight: Float, upperVariance: Float = 0f, lowerVariance: Float = upperVariance, image: String? = null) {
        this.key = key
        this.name = name
        this.weight = weight
        this.upperVariance = upperVariance
        this.lowerVariance = lowerVariance
        this.image = image
    }

    constructor(parcel: Parcel) {
        key = parcel.readString()
        name = parcel.readString()!!
        weight = parcel.readFloat()
        upperVariance = parcel.readFloat()
        lowerVariance = parcel.readFloat()
        image = parcel.readString()
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(key)
        dest.writeString(name)
        dest.writeFloat(weight)
        dest.writeFloat(upperVariance)
        dest.writeFloat(lowerVariance)
        dest.writeString(image)
    }

    companion object CREATOR : Creator<ItemImpl> {
        override fun createFromParcel(parcel: Parcel): ItemImpl {
            return ItemImpl(parcel)
        }

        override fun newArray(size: Int): Array<ItemImpl?> {
            return arrayOfNulls(size)
        }
    }
}